### DISCUSSION TITLE
Potential to relax TokenTypeIdentifier in schema to support more flexible permission/relation names?

### DISCUSSION THREAD

**Discussion by theronic** [authzed/spicedb#1904] [Q&A] [answered]
Would it be possible to relax the `TokenTypeIdentifier` pattern (`"^[a-z][a-z0-9_]{1,62}[a-z0-9]$"`) to support namespaced permission names that contain periods, slashes and maybe hyphens? E.g. `server.disks/create-disk`, `server/create` or `server/server$identifier`?

E.g.

```
definition server {
  relation owner: user
  permission server.disk/create = owner
}
```

In our Clojure application layer we would like to use fully-namespaced keywords as permission names such as `:server.disk/create` and encode them to/from valid Spice names. However, the limited alphanumeric + underscore pattern means we'll have to use multiple underscores for any kind of special encoding.

For example,  we can do something like`(encode :server.disks/create-disk) => "server__disk___create_disk"` but now the schema becomes error-prone and hard to write. We also need to deal with hyphen/underscore ambiguity, e.g. `(encode :servers.some-namespace/do_thing) => "server_disk__thing___do_thing")` which is not easily reversible without introducing even more underscores.

Understandably hyphens are tricky because they are used as minus relation logic operators.

There are workarounds like not using namespaces in our application, but it would be nice if our namespaced permissions matched the Spice schema more closely.

Would it be possible to support periods and slashes in the middle of permission name tokens without complicating the parser too much?

**Comment by josephschorr** [comment, marked as answer]
We'd have to investigate our various parsers to be 100% certain that adding `/` and `.` would not cause overlap or ambiguity, but I suspect it would be fine. We'll investigate more deeply and update here once we've double checked.

**Reply by theronic** [reply]
Thanks for your consideration :)

**Reply by vroldanbet** [reply]
@theronic feel free to open an issue
